### ISSUE TITLE
Potential issue with vehicle client-server desync (helicopters)

### ISSUE THREAD

**Issue by benvalkin** [SmartlyDressedGames/Unturned-3.x-Community#2906]
@SDGNelson
I am currently working on a set of custom helicopters, and upon testing them, I have noticed some small issues with their movement on servers. This is likely just something up wrong, but I've checked all possible causes quite thoroughly already with no success and so I thought it might be worth asking about this.

On my test server, it seems as though the helicopters' server-side clip is moving out of place with the client's vehicle.
I first noticed this after projectiles launched by the drivers' seat didn't explode where they were on the client.
I am aware that you said that driver's seat projectiles are currently unsupported here https://github.com/SmartlyDressedGames/Unturned-3.x-Community/issues/2455# but going off of the same fix from this thread (which was to add `Turret_0_Ignore_Camera_Aim` to the .dat file), I have been unable to achieve the same results. For me, the projectiles often explode at a completely different location to where they appeared to have landed on the client's screen, especially when the helicopter is angled downwards an moving quickly. For other players' in the server, they see rockets travel towards the position of the actual explosion. Here is a demo video:

https://user-images.githubusercontent.com/69928007/143768818-21dd3441-bfc2-4eb7-a7db-e553541cb52b.mp4

I also tested this with another passenger in the gunners' seat firing normal bullets. Obviously the location of the bullet impact was correct, since from what I understand regular bullet hits are registered on client and sent to the server, but the tracer particles often seemed to be travelling towards a completely different location from the impact point. I don't have a video of this, but I can record one if need be.

I then also noticed that if I made the vehicle travel significantly fast enough, the server would often try to correct its movement every few seconds as if the clip was frequently moving too far away from the client's vehicle. I'm not sure if this is just a security feature in place to prevent vehicles from moving too fast on servers, but it definitely makes me more suspcious.

Lastly, with my other (seemingly) working helicopters currently in use on my server, I've received some reports of projectiles disappearing after striking these vehicles in mid-air, especially while travelling fast. This leads me to believe that these helicopters have been suffering from server-side desync this whole time.

So again, I'm not sure if I'm setting them up wrongly. Here is what I've tried so far:
1) Making sure that the Vehicle and Clip rigidbody components had the same mass, drag, etc (they are completely identical)
2) Making sure that the Clip did not have any missing children components that could affect movement
3) Making sure that there were no other physics-based components on either gameobject that could affect the vehicles' movement
4) Building the vehicle off of the Skycrane example with as little changed possible

Any help or feedback is appreciated.

.

**Comment by SDGNelson** [comment]
Unfortunately it sounds like you are hitting the limits of the current vehicles. Since vehicles use the client physics results the server is always behind the client, but then projectiles are launched they use the server position. There are two related improvements I intend to make sooner or later, similar to the recent walking movement improvements:

1. Rewriting projectiles/throwables to use simpler deterministic-ish physics queries with better network prediction. This would fix the de-sync between client and server projectiles.

2. Experimenting with doing all the vehicle physics on the server. This would prevent vehicle cheats entirely and stop vehicle rubberbanding.

**Comment by benvalkin** [comment]
If you're interested, I actually managed to fix this this issue. Turns out it was a physics behaviour problem caused by having too many individual gameobjects each with their own single collider. This seems to make the vehicles' physics bug out and behave differently on client and server. The fix seems to have been to rather use one gameobject with multiple collider components.

So it seems like there's actually not too much wrong with the current implementation of client/server vehicle physics :)
